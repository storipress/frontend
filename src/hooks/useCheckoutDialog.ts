import { captureException } from '@sentry/vue'
import invariant from 'tiny-invariant'
import { GetBillingDocument, GetSiteCustomSiteDocument } from '~/graphql-operations'
import { RoleCode } from '~/utils/definition'

export const paidRoleList = [RoleCode.admin, RoleCode.editor]
export enum PlanStatus {
  Trial = 'Trial',
  Free = 'Free',
  Subscribed = 'Subscribed',
}

export type Mode = 'standard' | 'plus'
export type StripePlanGroup = 'blogger' | 'publisher' | 'prophet'
export type AppSumoPlanGroup = 'storipress_tier1' | 'storipress_tier2' | 'storipress_tier3'
export type AppSumoBfPlanGroup = 'storipress_bf_tier1' | 'storipress_bf_tier2' | 'storipress_bf_tier3'
export type AllPlanGroup = StripePlanGroup | AppSumoPlanGroup | AppSumoBfPlanGroup

const stripePlan = new Set<StripePlanGroup>(['blogger', 'publisher', 'prophet'])
const appSumoPlan = new Set<AppSumoPlanGroup>(['storipress_tier1', 'storipress_tier2', 'storipress_tier3'])
const appSumoBfPlan = new Set<AppSumoBfPlanGroup>(['storipress_bf_tier1', 'storipress_bf_tier2', 'storipress_bf_tier3'])

const allPlan = new Set<AllPlanGroup>([...stripePlan, ...appSumoPlan, ...appSumoBfPlan])
const plusPlan = new Set(['publisher', 'storipress_tier3', 'prophet'])

const isOpen = ref(false)
const checkoutMode = ref<Mode>('standard')
const checkoutFrom = ref('')
const checkoutDone = ref<Mode | null>()

export function useCheckoutDialog() {
  const { result: billingQueryResult, refetch: refetchBilling, loading: loadingBilling } = useQuery(GetBillingDocument)
  const { result: siteQueryResult, loading: loadingSite } = useQuery(GetSiteCustomSiteDocument)
  const billing = computed(() => billingQueryResult.value?.billing)
  const isAppSumo = computed(() => billing.value?.source === 'appsumo')
  function toAppSumo() {
    const appSumoPath = isAppSumo.value
      ? `https://appsumo.com/account/redemption/${billing.value?.plan_id || ''}#change-plan`
      : null
    invariant(appSumoPath, 'No AppSumo path')
    location.href = appSumoPath
  }

  watch(checkoutDone, () => refetchBilling())

  const openDialog = (mode: Mode = 'standard', from = '') => {
    if (isAppSumo.value) {
      toAppSumo()
      return false
    }

    checkoutMode.value = mode
    checkoutFrom.value = from
    if (billing.value?.subscribed && checkoutDone.value === checkoutMode.value) return true

    isOpen.value = true
    return Promise.any([until(checkoutDone).toBeTruthy(), until(isOpen).not.toBeTruthy()])
  }
  const closeDialog = () => (isOpen.value = false)
  const doneCheckout = () => {
    checkoutDone.value = checkoutMode.value
    isOpen.value = false
  }

  const planStatus = computed(() => {
    const plan = siteQueryResult.value?.site.plan as AllPlanGroup | undefined
    if (!plan) return PlanStatus.Free
    if (allPlan.has(plan)) {
      return PlanStatus.Subscribed
    }
    if ((plan as string) !== 'free') {
      captureException(new Error(`found unknown plan: ${plan}`))
    }
    return PlanStatus.Free
  })

  const groupKey = computed(() => {
    const modeGroupMap: Record<Mode, AllPlanGroup> = {
      standard: 'blogger',
      plus: 'publisher',
    }
    return modeGroupMap[checkoutMode.value]
  })

  const isPlusPlan = computed(() => {
    const plan = siteQueryResult.value?.site.plan as AllPlanGroup
    return plusPlan.has(plan)
  })

  const isFreePlan = computed(() => {
    return planStatus.value === PlanStatus.Free
  })

  const loading = computed(() => loadingBilling.value || loadingSite.value)

  async function requirePlusPlan(source: string) {
    if (isPlusPlan.value) {
      return true
    }

    return await openDialog('plus', source)
  }

  return {
    isOpen,
    checkoutMode,
    checkoutFrom,
    billing,
    planStatus,
    loadingBilling,
    loading,
    groupKey,
    isPlusPlan,
    isFreePlan,
    refetchBilling,
    openDialog,
    closeDialog,
    doneCheckout,
    checkPaidRole,
    requirePlusPlan,
  }
}

export default useCheckoutDialog

export function checkPaidRole(roleCode: number | string) {
  return paidRoleList.includes(Number(roleCode))
}
