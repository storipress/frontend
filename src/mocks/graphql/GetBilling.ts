import { defineGraphQLMock } from '../define-graphql-mock'
import { GetBillingDocument } from '~/graphql-operations'

export default defineGraphQLMock(GetBillingDocument, {
  billing: {
    id: '1',
    on_trial: false,
    canceled: false,
    credit_balance: '0',
    has_historical_subscriptions: false,
    has_pm: true,
    next_pm_discounts: [],
    next_pm_taxes: [],
    on_grace_period: false,
    publications_count: 1,
    publications_quota: 99,
    seats_in_use: 1,
    subscribed: true,
    next_pm_date: '2020-01-01',
    trial_ends_at: null,
    ends_at: null,
    source: 'stripe',
    interval: 'monthly',
    next_pm_subtotal: '0',
    next_pm_tax: '0',
    next_pm_total: '0',
    plan: 'publisher',
    plan_id: '1',
    pm_last_four: '0000',
    pm_type: 'visa',
    quantity: 1,
    referer: null,
    has_prophet: false,
  },
})
