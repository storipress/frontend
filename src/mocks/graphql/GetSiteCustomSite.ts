import { defineGraphQLMock } from '../define-graphql-mock'
import { GetSiteCustomSiteDocument } from '~/graphql-operations'

export default defineGraphQLMock(GetSiteCustomSiteDocument, {
  site: {
    id: 'D6RX98VXN',
    name: "Jack's Blog",
    plan: 'publisher',
    custom_site_template: false,
    newstand_key: 'newstand_key',
    typesense_search_only_key: 'typesense_search_only_key',
  },
})
