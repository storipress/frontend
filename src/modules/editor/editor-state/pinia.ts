import { reactive } from 'vue'
import { defineStore } from 'pinia'
import type { Client } from '~/components/Editor/core/types'

interface HistoryState {
  undo: boolean
  redo: boolean
}

const initialState = {
  history: {
    undo: false,
    redo: false,
  },
  saved: true,
  html: '',
  plaintext: '',
  scrollTop: 0,
  editorTop: 0,
  clients: [] as Client[],
}

export const editorState = defineStore('editorState', () => {
  const state = reactive(initialState)

  return {
    state,
    SET_HISTORY(history: HistoryState) {
      state.history = history
    },

    SET_SAVED(saved: boolean) {
      state.saved = saved
    },

    SET_HTML(html: string) {
      state.html = html
    },

    SET_PLAINTEXT(plaintext: string) {
      state.plaintext = plaintext
    },

    SET_CLIENTS(clients: Client[]) {
      state.clients = clients
    },

    SET_SCROLL_TOP(scrollTop: number) {
      state.scrollTop = scrollTop
    },

    SET_EDITOR_TOP(editorTop: number) {
      state.editorTop = editorTop
    },
  }
})
