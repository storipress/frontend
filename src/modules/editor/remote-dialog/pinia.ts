import { ref } from 'vue'
import { defineStore } from 'pinia'
import type { DialogInfo } from './types'

export const remoteDialog = defineStore(
  'remoteDialog',
  () => {
    const info = ref<DialogInfo<string> | null>(null)

    return {
      info,
      SET_DIALOG(newInfo: DialogInfo | null) {
        info.value = newInfo
      },

      SET_RETURN(newReturnValue: unknown) {
        if (!info.value) return
        info.value.returnValue = newReturnValue
      },
    }
  },
  { sync: { enabled: true } },
)
