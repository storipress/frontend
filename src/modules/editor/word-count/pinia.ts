import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import { EMPTY_STATISTICS } from './constants'
import { getImageReadSeconds, getTotalReadMinute, getWordReadSeconds } from './read-time'
import type { WordStatistics } from './types'

export interface State {
  statistics: WordStatistics
}

export const useWordCount = defineStore('wordCount', () => {
  const statistics = ref(EMPTY_STATISTICS)

  const imageReadTime = computed(() => {
    return getImageReadSeconds(statistics.value.images)
  })

  const readTime = computed(() => {
    const wordSeconds = getWordReadSeconds(statistics.value.words)
    return getTotalReadMinute(wordSeconds, imageReadTime.value)
  })

  return {
    statistics,
    imageReadTime,
    readTime,

    SET_STATISTICS(stat: WordStatistics) {
      statistics.value = stat
    },
  }
})
