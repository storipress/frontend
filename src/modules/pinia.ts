import { createPinia } from 'pinia'
import { sync } from '~/lib/sync'
import type { UserModule } from '~/types'

// Safety: we are not using SSR
export const pinia = createPinia()

// Setup Pinia
// https://pinia.esm.dev/
export const install: UserModule = ({ isClient, initialState, app }) => {
  pinia.use(sync)
  app.use(pinia)
  // Refer to
  // https://github.com/antfu/vite-ssg/blob/main/README.md#state-serialization
  // for other serialization strategies.
  if (isClient) pinia.state.value = initialState.pinia || {}
  else initialState.pinia = pinia.state.value
}
