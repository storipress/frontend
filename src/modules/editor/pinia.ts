import { Evt } from 'evt'
import { onUnmounted } from 'vue'
import { editorState as useEditorState } from './editor-state/pinia'
import { useWordCount } from './word-count/pinia'
import type { Client, WordStatistics } from '~/components/Editor/core/types'
import { clients, historyState, html, plaintext, saved, wordCount } from '~/utils/editor/global-bus'

export { useEditorState }
export { remoteDialog as useRemoteDialog } from './remote-dialog/pinia'
export { useWordCount } from './word-count/pinia'

export function useListenEditorState() {
  const editorState = useEditorState()
  const wordCountStore = useWordCount()

  function handleHistory(history: { undo: boolean; redo: boolean }) {
    editorState.SET_HISTORY(history)
  }

  function handleSaved(saved: boolean) {
    if (!saved) {
      editorState.SET_SAVED(false)
    }
  }

  function handleWordCount(statistics: WordStatistics) {
    wordCountStore.SET_STATISTICS(statistics)
  }

  function handlePlainText(plaintext: string) {
    editorState.SET_PLAINTEXT(plaintext)
  }

  function handleClients(clients: Client[]) {
    editorState.SET_CLIENTS(clients)
  }

  function handleHTML(html: string) {
    editorState.SET_HTML(html)
  }

  handleHistory(historyState.state)
  handleSaved(saved.state)
  handleWordCount(wordCount.state)
  handlePlainText(plaintext.state)
  handleClients(clients.state)
  handleHTML(html.state)

  const ctx = Evt.newCtx()

  historyState.attach(ctx, handleHistory)
  saved.attach(ctx, handleSaved)
  wordCount.attach(ctx, handleWordCount)
  plaintext.attach(ctx, handlePlainText)
  clients.attach(ctx, handleClients)
  html.attach(ctx, handleHTML)

  onUnmounted(() => {
    ctx.done()
  })
}
