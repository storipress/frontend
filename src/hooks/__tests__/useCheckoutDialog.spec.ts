import { setActivePinia } from 'pinia'
import { captureException } from '@sentry/vue'
import { PlanStatus, useCheckoutDialog } from '../useCheckoutDialog'
import { mockResponseOnce, setupApolloClient, setupTestPinia } from '~/test-helpers'
import GetBilling from '~/mocks/graphql/GetBilling'
import GetSiteCustomSite from '~/mocks/graphql/GetSiteCustomSite'

vi.mock('@sentry/vue', async () => ({
  ...(await vi.importActual('@sentry/vue')),
  captureException: vi.fn(),
}))

beforeEach(() => {
  setActivePinia(setupTestPinia())
  setupApolloClient()
})

it('should return checkout info', async () => {
  const checkoutDialog = useCheckoutDialog()

  await until(() => checkoutDialog.loading.value).toBe(false)
  await until(() => checkoutDialog.billing.value).toBeTruthy()

  expect(reactive(checkoutDialog)).toMatchSnapshot()
})

// FIXME: this test can pass if we add `.only` on it, seems it's a cache issue
it.skip('should warn for unknown plan', async () => {
  mockResponseOnce(GetSiteCustomSite, {
    site: {
      plan: 'unknown',
    },
  })
  mockResponseOnce(GetBilling, {
    billing: {
      plan: 'unknown',
    },
  })
  const checkoutDialog = useCheckoutDialog()

  await until(() => checkoutDialog.loading.value).toBe(false)
  await until(() => checkoutDialog.billing.value).toBeTruthy()

  expect(checkoutDialog.planStatus.value).toBe(PlanStatus.Free)

  expect(captureException).toHaveBeenCalledTimes(1)
  expect(captureException).toHaveBeenCalledWith(new Error('found unknown plan: unknown'))
  expect(reactive(checkoutDialog)).toMatchSnapshot()
})
