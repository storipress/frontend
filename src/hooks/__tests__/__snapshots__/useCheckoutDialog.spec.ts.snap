// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should return checkout info 1`] = `
{
  "billing": {
    "canceled": false,
    "credit_balance": "0",
    "ends_at": null,
    "has_historical_subscriptions": false,
    "has_pm": true,
    "id": "1",
    "interval": "monthly",
    "next_pm_date": "2020-01-01",
    "next_pm_discounts": [],
    "next_pm_subtotal": "0",
    "next_pm_tax": "0",
    "next_pm_taxes": [],
    "next_pm_total": "0",
    "on_grace_period": false,
    "on_trial": false,
    "plan": "publisher",
    "plan_id": "1",
    "pm_last_four": "0000",
    "pm_type": "visa",
    "publications_count": 1,
    "publications_quota": 99,
    "quantity": 1,
    "referer": null,
    "seats_in_use": 1,
    "source": "stripe",
    "subscribed": true,
    "trial_ends_at": null,
  },
  "checkPaidRole": [Function],
  "checkoutFrom": "",
  "checkoutMode": "standard",
  "closeDialog": [Function],
  "doneCheckout": [Function],
  "groupKey": "blogger",
  "isFreePlan": false,
  "isOpen": false,
  "isPlusPlan": true,
  "loading": false,
  "loadingBilling": false,
  "openDialog": [Function],
  "planStatus": "Subscribed",
  "refetchBilling": [Function],
  "requirePlusPlan": [Function],
}
`;

exports[`should warn for unknown plan 1`] = `
{
  "billing": {
    "canceled": false,
    "credit_balance": "0",
    "ends_at": null,
    "has_historical_subscriptions": false,
    "has_pm": true,
    "id": "1",
    "interval": "monthly",
    "next_pm_date": "2020-01-01",
    "next_pm_discounts": [],
    "next_pm_subtotal": "0",
    "next_pm_tax": "0",
    "next_pm_taxes": [],
    "next_pm_total": "0",
    "on_grace_period": false,
    "on_trial": false,
    "plan": "unknown",
    "plan_id": "1",
    "pm_last_four": "0000",
    "pm_type": "visa",
    "publications_count": 1,
    "publications_quota": 99,
    "quantity": 1,
    "referer": null,
    "seats_in_use": 1,
    "source": "stripe",
    "subscribed": true,
    "trial_ends_at": null,
  },
  "checkPaidRole": [Function],
  "checkoutFrom": "",
  "checkoutMode": "standard",
  "closeDialog": [Function],
  "doneCheckout": [Function],
  "groupKey": "blogger",
  "isFreePlan": true,
  "isOpen": false,
  "isPlusPlan": false,
  "loading": false,
  "loadingBilling": false,
  "openDialog": [Function],
  "planStatus": "Free",
  "refetchBilling": [Function],
}
`;
